RecoTauTag/HLTProducers/interface/L1TJetsMatching.h

```diff
@@ -0,0 +1,210 @@
+
+#include "Math/GenVector/VectorUtil.h"
+#include "DataFormats/HLTReco/interface/TriggerTypeDefs.h"
+#include "FWCore/Utilities/interface/EDMException.h"
+#include "DataFormats/JetReco/interface/PFJet.h"
+
+#include "HLTrigger/HLTcore/interface/defaultModuleLabel.h"
+#include "DataFormats/Math/interface/deltaR.h"
+
+#include <map>
+#include <vector>
+
+template< typename T>
+class L1TJetsMatching: public edm::global::EDProducer<> {
+ public:
+  explicit L1TJetsMatching(const edm::ParameterSet&);
+  ~L1TJetsMatching();
+  virtual void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+ private:
+    
+  const edm::EDGetTokenT<std::vector<T>> jetSrc_;
+  const edm::EDGetTokenT<trigger::TriggerFilterObjectWithRefs> jetTrigger_;
+  const double pt1Min_;
+  const double pt2Min_;
+  const double mjjMin_;
+  const double matchingR_;
+  const double matchingR2_;
+  };
+    //
+    // class decleration
+    //
+    using namespace reco   ;
+    using namespace std    ;
+    using namespace edm    ;
+    using namespace trigger;
+    
+    
+    template< typename T>
+    std::pair<std::vector<T>,std::vector<T>> categorise(const std::vector<T>& pfMatchedJets, double pt1, double pt2, double Mjj)
+    {
+        std::pair<std::vector<T>,std::vector<T>> output;
+        unsigned int i1 = 0;
+        unsigned int i2 = 0;
+        double mjj = 0;
+        if (pfMatchedJets.size()>1){
+            for (unsigned int i = 0; i < pfMatchedJets.size()-1; i++){
+                
+                const T &  myJet1 = (pfMatchedJets)[i];
+                
+                for (unsigned int j = i+1; j < pfMatchedJets.size(); j++)
+                {
+                    const T &  myJet2 = (pfMatchedJets)[j];
+                    
+                    const double mjj_test = (myJet1.p4()+myJet2.p4()).M();
+                    
+                    if (mjj_test > mjj){
+                        
+                        mjj =mjj_test;
+                        i1 = i;
+                        i2 = j;
+                    }
+                }
+            }
+            
+            const T &  myJet1 = (pfMatchedJets)[i1];
+            const T &  myJet2 = (pfMatchedJets)[i2];
+            
+            if ((mjj > Mjj) && (myJet1.pt() >= pt1) && (myJet2.pt() > pt2) )
+            {
+                
+                output.first.push_back(myJet1);
+                output.first.push_back(myJet2);
+                
+            }
+            
+            if ((mjj > Mjj) && (myJet1.pt() < pt1) && (myJet1.pt() > pt2) && (myJet2.pt() > pt2))
+            {
+                
+                const T &  myJetTest = (pfMatchedJets)[0];
+                if (myJetTest.pt()>pt1){
+                    output.second.push_back(myJet1);
+                    output.second.push_back(myJet2);
+                    output.second.push_back(myJetTest);
```

Comment:
i2 should not be able to be ==0 by construction (j=i+1,...). Also, I think that it is safe even without i1==0 protection, because of the pt>pt1 requirement (in the if above there is a req for myJet1.pt<=pt1 so if i1 = 0 it will fail the statement at the end)